notification/slack/messageoptions.go

```diff
@@ -0,0 +1,95 @@
+package slack
+
+import (
+	"strconv"
+
+	"github.com/slack-go/slack"
+	"github.com/target/goalert/alert"
+)
+
+func alertIDAndStatusSection(alertID string, status alert.Status) *slack.HeaderBlock {
+	var s string
+	switch status {
+	case alert.StatusTriggered:
+		s = "Unacknowledged"
+	case alert.StatusActive:
+		s = "Acknowledged"
+	case alert.StatusClosed:
+		s = "Closed"
+	default:
+		panic("alert type not supported")
+	}
+	txt := alertID + ": " + s
+	summaryText := slack.NewTextBlockObject("plain_text", txt, false, false)
+	return slack.NewHeaderBlock(summaryText)
+}
+
+func alertSummarySection(summary string) *slack.SectionBlock {
+	summaryText := slack.NewTextBlockObject("mrkdwn", "Summary: "+summary, false, false)
+	return slack.NewSectionBlock(summaryText, nil, nil)
+}
+
+func ackButton(alertID string) slack.ButtonBlockElement {
+	txt := slack.NewTextBlockObject("plain_text", "Acknowledge :eyes:", true, false)
+	return *slack.NewButtonBlockElement("ack", alertID, txt)
+}
+
+func escButton(alertID string) *slack.ButtonBlockElement {
+	txt := slack.NewTextBlockObject("plain_text", "Escalate :arrow_up:", true, false)
+	return slack.NewButtonBlockElement("esc", alertID, txt)
+}
+
+func closeButton(alertID string) *slack.ButtonBlockElement {
+	txt := slack.NewTextBlockObject("plain_text", "Close :ballot_box_with_check:", true, false)
+	return slack.NewButtonBlockElement("close", alertID, txt)
+}
+
+func openLinkButton(url string) *slack.ButtonBlockElement {
+	txt := slack.NewTextBlockObject("plain_text", "Open in GoAlert :link:", true, false)
+	s := slack.NewButtonBlockElement("openLink", "", txt)
+	s.URL = url
+	return s
+}
```

Comment:
This seems really noisy, can we make it a bit more generic instead of a bunch of 2 line functions?
Here's an example of what I'm thinking:

```go
// one helper func

type buttonType int

const (
	buttonTypeAck buttonType = iota
	buttonTypeEsc
	buttonTypeClose
	buttonTypeOpenLink
)

func makeButtons(callbackID, url string) func(buttonType) *slack.ButtonBlockElement {
	return func(typ buttonType) *slack.ButtonBlockElement {
		var action, text, setURL string
		switch typ {
		case buttonTypeAck:
			action = "ack"
			text = "Acknowledge :eyes:"
		case buttonTypeEsc:
			action = "esc"
			text = "Escalate :arrow_up:"
		case buttonTypeClose:
			action = "close"
			text = "Close :ballot_box_with_check:"
		case buttonTypeOpenLink:
			action = "openLink"
			text = "Open in GoAlert :link:"
			setURL = url
		default:
			panic("button type not supported")
		}

		el := slack.NewButtonBlockElement(
			action,
			callbackID,
			slack.NewTextBlockObject("plain_text", text, true, false),
		)
		if setURL != "" {
			el.URL = setURL
		}
		return el
	}
}

// and then inside the render func
	button := makeButtons(callbackID, url)
	switch a.Status {
	case alert.StatusTriggered:
		actions = slack.NewActionBlock("",
			button(buttonTypeAck),
			button(buttonTypeEsc),
			button(buttonTypeClose),
			button(buttonTypeOpenLink),
		)
	case alert.StatusActive:
		actions = slack.NewActionBlock("",
			button(buttonTypeEsc),
			button(buttonTypeClose),
			button(buttonTypeOpenLink),
		)
	case alert.StatusClosed:
		actions = slack.NewActionBlock("",
			button(buttonTypeOpenLink),
		)
	}
```
---
Two switches with only the relevant diffs -- helper method just switches the text and actions of the buttons, and the render method just switches which buttons we're including in the message. A type enum makes it clear what button types we're supporting/using.

Reply:
yeah I can definitely give that a shot